sn_interface/src/network_knowledge/sections_dag.rs

```diff
@@ -0,0 +1,1020 @@
+    pub fn partial_dag(&self, from: &bls::PublicKey, to: &bls::PublicKey) -> Result<Self> {
+        // start from the "to" key (bottom of the tree) and traverse to the root
+        let mut crdt_ops: Vec<Node<SectionInfo>> = Vec::new();
+        // if "to" key is genesis, returns error
+        let mut node = self.get_node(to)?;
+        loop {
+            if node.value.key == *from {
+                // we have reached the end
+                break;
+            }
+            // we don't insert the `from` node
+            crdt_ops.push(node.clone());
+            match self.parent_node(node.hash()) {
+                Some(parent) => node = parent,
+                // reached root of the dag
+                None => {
+                    // after reaching dag_root, return the partial_dag if from == genesis
+                    if *from == self.genesis_key {
+                        break;
+                    }
+                    // from key was apparently not a direct ancestor
+                    return Err(Error::InvalidBranch);
+                }
+            }
+        }
+        // create the SectionsDAG and apply the CRDT ops
+        let mut dag = SectionsDAG::new(*from);
+        let mut parent = *from;
+        for node in crdt_ops.into_iter().rev() {
+            let key = node.value.key;
+            dag.insert_node(&parent, node)?;
+            parent = key;
+        }
+        Ok(dag)
+    }
+
+    /// Update our current `SectionsDAG` with the keys from another `SectionsDAG`
+    /// Returns `Error::InvalidSignature` if the provided DAG fails signature verification
+    /// Returns `Error::KeyNotFound` if the genesis_key of either of the DAGs is not present in the
+    /// other
+    pub fn join(&mut self, mut other: Self) -> Result<()> {
+        if !other.self_verify() {
+            return Err(Error::InvalidSignature);
+        }
+        // find which DAG is the parent
+        if !self.keys().contains(other.genesis_key()) {
+            if other.keys().contains(self.genesis_key()) {
+                mem::swap(self, &mut other);
+            } else {
+                return Err(Error::KeyNotFound(self.genesis_key));
+            }
+        }
+        if other.dag_root.is_empty() {
+            return Ok(());
+        }
+        // insert all the nodes from `other`
+        let mut need_to_visit: Vec<Node<SectionInfo>> = Vec::new();
+        for key in other.dag_root.iter() {
+            need_to_visit.push(other.get_node(key)?);
+        }
+        while let Some(node) = need_to_visit.pop() {
+            // cannot be None since we don't provide the genesis_key
+            let parent = match other.get_parent_key(&node.value.key)? {
+                Some(parent) => parent,
+                None => {
+                    panic!("SectionsDAG::join() Unreachable since key can never be `genesis_key`")
+                }
+            };
+            need_to_visit.extend(other.child_nodes(node.hash()).into_iter());
```

Comment:
vec.extend() takes an `IntoIterator` so you don't need to call `into_iter()`

```suggestion
            need_to_visit.extend(other.child_nodes(node.hash()));
```